swig/cpp/src/Session.cpp

```
@@ -0,0 +1,281 @@
+    case(SR_ERR_INIT_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_INIT_FAILED));
+    case(SR_ERR_IO):
+        throw runtime_error(sr_strerror(SR_ERR_IO));
+    case(SR_ERR_DISCONNECT):
+        throw runtime_error(sr_strerror(SR_ERR_DISCONNECT));
+    case(SR_ERR_MALFORMED_MSG):
+        throw runtime_error(sr_strerror(SR_ERR_MALFORMED_MSG));
+    case(SR_ERR_UNSUPPORTED):
+        throw runtime_error(sr_strerror(SR_ERR_UNSUPPORTED));
+    case(SR_ERR_UNKNOWN_MODEL):
+        throw runtime_error(sr_strerror(SR_ERR_UNKNOWN_MODEL));
+    case(SR_ERR_BAD_ELEMENT):
+        throw runtime_error(sr_strerror(SR_ERR_BAD_ELEMENT));
+    case(SR_ERR_VALIDATION_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_VALIDATION_FAILED));
+    case(SR_ERR_COMMIT_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_COMMIT_FAILED));
+    case(SR_ERR_DATA_EXISTS):
+        throw runtime_error(sr_strerror(SR_ERR_DATA_EXISTS));
+    case(SR_ERR_DATA_MISSING):
+        throw runtime_error(sr_strerror(SR_ERR_DATA_MISSING));
+    case(SR_ERR_UNAUTHORIZED):
+        throw runtime_error(sr_strerror(SR_ERR_UNAUTHORIZED));
+    case(SR_ERR_LOCKED):
+        throw runtime_error(sr_strerror(SR_ERR_LOCKED));
+    case(SR_ERR_TIME_OUT):
+        throw runtime_error(sr_strerror(SR_ERR_TIME_OUT));
+    }
+}
+
+Session::Session(Connection& sys, const sr_conn_options_t opts, const char *user_name)
+{
+    int rc;
+    _opts = opts;
+    _datastore = SR_DS_STARTUP;
+
+    if (user_name == NULL) {
+        /* start session */
+        rc = sr_session_start(sys.get_conn(), _datastore, _opts, &_sess);
+        if (SR_ERR_OK != rc) {
+            goto cleanup;
+        }
+    } else {
+        /* start session */
+        rc = sr_session_start_user(sys.get_conn(), user_name, _datastore, _opts, &_sess);
+        if (SR_ERR_OK != rc) {
+            goto cleanup;
+        }
+    }
+
+    return;
+
+cleanup:
+    throw_excption(SR_ERR_INIT_FAILED);
```

You should rather throw the exception matching with error stored in rc.